### ISSUE TITLE
Enable trigger by registry

### ISSUE THREAD

**Issue by samcro1967** [getwud/wud#161]
Is there a way to enable a trigger in docker/docker compose for all images/containers for a particular registry?

Trying to manage the Docker Hub rate limiting and minimize the labels in docker compose.  I would like to have `WUD_WATCHER_LOCAL_WATCHBYDEFAULT=false` set that disables all and then have a trigger like `WUD_REGISTRY_GHCR_WATCHBYDEFAULTOVERRIDE=true`  that enables it for everything in this registry.  Then I can use `wud.watch=true` to enable it for only specific images from Docker Hub.

I don't think there is a way to currently do this looking through the docs, but maybe I missed it.

**Comment by fmartinou** [comment]
It's not possible right now.

Other issues seem pretty similar.
Like https://github.com/fmartinou/whats-up-docker/issues/133

I have to think about...

**Comment by samcro1967** [comment]
I have been able to eliminate regularly hitting the docker hub rate limit by setting up a second local docker registry as a pull through cache.  Using less than 10 of my allocations every 6 hours now with whatsupdocker running every hour.  Thought I would let you know in case you wanted to mention it in the readme.  Guessing others may be bumping up against it as well and may find it useful.

**Comment by fmartinou** [comment]
Nice workaround 👍 
If you can provide an example of the configuration to set up, I'll add it somewhere to the documentation.
Thank you!

**Comment by samcro1967** [comment]
My documentation is kind of messy, but the steps are not complicated.  I was able to get this working and tested in about 20 minutes.  Hope this helps others.

Also worth noting that a local docker registry can not be used as both a push and pull through cache.  So I am running 2 instances.  One is for my local CI/CD pipeline to push new images to and the other is just for the local cache.

**Confirm current rate limit**
Look at ratelimit-remaining in the output
```
TOKEN=$(curl "https://auth.docker.io/token?service=registry.docker.io&scope=repository:ratelimitpreview/test:pull" | jq -r .token) && curl --head -H "Authorization: Bearer $TOKEN" https://registry-1.docker.io/v2/ratelimitpreview/test/manifests/latest
```

1. Delete a container and its image.  Pull it again.  Rerun the above command to confirm the count decrements by 1.
2. Install the mirror using the instruction below.
3. Delete a container and its image.  Pull it again.  Rerun the above command to confirm the count decrements by 1.  This is because the image is not cached locally yet.
4. Delete the same container and its image.  Pull it again.  Rerun the above command to confirm the count remains the same as it is now cached locally.  Every subsequent pull will go to the cache first.  The cache will check docker hub to see if a newer version is available and if so will pull it.  The count will only decrement if the image is newer and needs to be pulled.  If it is not newer and the pull is served by the local cache, the count will not decrement.
5. You can see what images are cached locally looking at `/var/lib/registry/v2/repositories` in the container or from where ever you mounted the persistent volume on the host.

<details>

**<summary>Registry mirror docker compose</summary>**

```
  registry_mirror:
    hostname: registry_mirror
    image: registry:latest
    container_name: registry_mirror
    restart: always
    environment:
      - REGISTRY_STORAGE_DELETE_ENABLED=true
    ports:
      - 5002:5000
    volumes:
      - ${docker}/registry_mirror/data:/var/lib/registry
      - ${docker}/registry_mirror/config:/etc/docker/registry
      - /etc/timezone:/etc/timezone:ro
      - /etc/localtime:/etc/localtime:ro
```

</details>

<details>

**<summary>Registry Mirror /etc/docker/registry/config.yml</summary>**
Note the last 2 lines of the config file
```
version: 0.1
log:
  fields:
    service: registry
storage:
  delete:
    enabled: true
  cache:
    blobdescriptor: inmemory
  filesystem:
    rootdirectory: /var/lib/registry
  maintenance:
    uploadpurging:
      enabled: true
      age: 24h
      interval: 24h
      dryrun: false
    readonly:
      enabled: false
http:
  addr: :5000
  headers:
    X-Content-Type-Options: [nosniff]
    Access-Control-Allow-Origin: ['http://osu.registry-ui:5001']
    Access-Control-Allow-Methods: ['HEAD', 'GET', 'OPTIONS', 'DELETE']
    Access-Control-Expose-Headers: ['Docker-Content-Digest']
health:
  storagedriver:
    enabled: true
    interval: 10s
    threshold: 3
proxy:
  remoteurl: https://registry-1.docker.io
```

</details>

**docker daemon.json**
```
{
        "registry-mirrors": ["http://osu.registry_mirror:5002"]
}
```
**Reload docker daemon**
```
sudo systemctl docker reload
```

**Verify mirror is loaded** 
Should see registry mirror located at the end of the output
```
docker system info
```

**Comment by samcro1967** [comment]
I seem to have spoken too soon.  The local registry cache is keeping the rate limit from decrementing when a docker pull is manually executed (both from the host and inside the container, but when the check runs in the container, the count is still being decremented.  When I check the limit from inside the container (and from the host) it shows I still have plenty of pulls left.  But when the check runs and I check the container logs, it says I have hit the pull limit.  Not sure what is going on at this point.

**Comment by samcro1967** [comment]
Guessing the code would need to be extended to recognize the presence of a local docker registry cache and when enable performs pulls against it rather than docker hub.

**Comment by samcro1967** [comment]
@fmartinou Struggling to understand why I am seeing this in the wud logs.  When I check my rate limit, there are plenty of pulls left.  I am authenticated to docker hub from my host and in wud as far as I can tell.  It shows the registry is loaded in the logs and authenticated with the user ID and token I provided.  There is no realy way for me to determine if it is I guess when the check runs.  I suspect there may be a bug here.


```
12:31:05.885  WARN whats-up-docker/watcher.docker.local: Error when processing (429 - {"errors":[{"code":"TOOMANYREQUESTS","message":"You have reached your pull rate limit. You may increase the limit by authenticating and upgrading: https://www.docker.com/increase-rate-limit"}]}) (container=local_glances)
```

```
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  4369    0  4369    0     0  19170      0 --:--:-- --:--:-- --:--:-- 19246
HTTP/1.1 200 OK
content-length: 2782
content-type: application/vnd.docker.distribution.manifest.v1+prettyjws
docker-content-digest: sha256:767a3815c34823b355bed31760d5fa3daca0aec2ce15b217c9cd83229e0e2020
docker-distribution-api-version: registry/2.0
etag: "sha256:767a3815c34823b355bed31760d5fa3daca0aec2ce15b217c9cd83229e0e2020"
date: Sat, 17 Sep 2022 17:31:47 GMT
strict-transport-security: max-age=31536000
ratelimit-limit: 100;w=21600
ratelimit-remaining: 97;w=21600
docker-ratelimit-source: xxx.xxx.xxx.xxx
```

**Comment by fmartinou** [comment]
Now it's possible to associate different triggers per container?
Would it be a viable solution for the initial issue?

(available from `7.2.0`)

**Comment by samcro1967** [comment]
I would say yes as it does give the ability to set triggers per container, but the original request is to set triggers by registry.  It looks like the same could be accomplished by putting labels on every container rather than just a label on the wud container itself.  There would also need to be a configurable cron trigger that could be applied per container to set different schedules per container on when to do a check.
